File: packages/network/src/v2/schemas/decodeStaticField.ts
import {
  getStaticByteLength,
  SchemaType,
  SchemaTypeToPrimitiveType,
  StaticSchemaType,
} from "@latticexyz/schema-type/deprecated";
import { toHex, pad } from "viem";

const unsupportedStaticField = (fieldType: never): never => {
  throw new Error(`Unsupported static field type: ${SchemaType[fieldType] ?? fieldType}`);
};

export const decodeStaticField = <T extends StaticSchemaType, P extends SchemaTypeToPrimitiveType<T>>(
  fieldType: T,
  bytes: Uint8Array,
  offset: number
): P => {
  const staticLength = getStaticByteLength(fieldType);
  const slice = bytes.slice(offset, offset + staticLength);
  const hex = toHex(slice);
  const numberHex = hex.replace(/^0x$/, "0x0");

  switch (fieldType) {
    case SchemaType.BOOL:
      return (Number(numberHex) !== 0) as P;
    case SchemaType.UINT8:
    case SchemaType.UINT16:
    case SchemaType.UINT24:
    case SchemaType.UINT32:
    case SchemaType.UINT40:
    case SchemaType.UINT48:
      return Number(numberHex) as P;
    case SchemaType.UINT56:
    case SchemaType.UINT64:
    case SchemaType.UINT72:
    case SchemaType.UINT80:
    case SchemaType.UINT88:
    case SchemaType.UINT96:
    case SchemaType.UINT104:
    case SchemaType.UINT112:
    case SchemaType.UINT120:
    case SchemaType.UINT128:
    case SchemaType.UINT136:
    case SchemaType.UINT144:
    case SchemaType.UINT152:
    case SchemaType.UINT160:
    case SchemaType.UINT168:
    case SchemaType.UINT176:
    case SchemaType.UINT184:
    case SchemaType.UINT192:
    case SchemaType.UINT200:
    case SchemaType.UINT208:
    case SchemaType.UINT216:
    case SchemaType.UINT224:
    case SchemaType.UINT232:
    case SchemaType.UINT240:
    case SchemaType.UINT248:
    case SchemaType.UINT256:
      return BigInt(numberHex) as P;
    case SchemaType.INT8:
    case SchemaType.INT16:
    case SchemaType.INT24:
    case SchemaType.INT32:
    case SchemaType.INT40:
    case SchemaType.INT48: {
      const max = 2 ** (staticLength * 8);
      const num = Number(numberHex);
      return (num < max / 2 ? num : num - max) as P;
    }
    case SchemaType.INT56:
    case SchemaType.INT64:
    case SchemaType.INT72:
    case SchemaType.INT80:
    case SchemaType.INT88:
    case SchemaType.INT96:
    case SchemaType.INT104:
    case SchemaType.INT112:
    case SchemaType.INT120:
    case SchemaType.INT128:
    case SchemaType.INT136:
    case SchemaType.INT144:
    case SchemaType.INT152:
    case SchemaType.INT160:
    case SchemaType.INT168:
    case SchemaType.INT176:
    case SchemaType.INT184:
    case SchemaType.INT192:
    case SchemaType.INT200:
    case SchemaType.INT208:
    case SchemaType.INT216:
    case SchemaType.INT224:
    case SchemaType.INT232:
    case SchemaType.INT240:
    case SchemaType.INT248:
    case SchemaType.INT256: {
      const max = 2n ** (BigInt(staticLength) * 8n);
      const num = BigInt(numberHex);
      return (num < max / 2n ? num : num - max) as P;
    }
    case SchemaType.BYTES1:
    case SchemaType.BYTES2:
    case SchemaType.BYTES3:
    case SchemaType.BYTES4:
    case SchemaType.BYTES5:
    case SchemaType.BYTES6:
    case SchemaType.BYTES7:
    case SchemaType.BYTES8:
    case SchemaType.BYTES9:
    case SchemaType.BYTES10:
    case SchemaType.BYTES11:
    case SchemaType.BYTES12:
    case SchemaType.BYTES13:
    case SchemaType.BYTES14:
    case SchemaType.BYTES15:
    case SchemaType.BYTES16:
    case SchemaType.BYTES17:
    case SchemaType.BYTES18:
    case SchemaType.BYTES19:
    case SchemaType.BYTES20:
    case SchemaType.BYTES21:
    case SchemaType.BYTES22:
    case SchemaType.BYTES23:
    case SchemaType.BYTES24:
    case SchemaType.BYTES25:
    case SchemaType.BYTES26:
    case SchemaType.BYTES27:
    case SchemaType.BYTES28:
    case SchemaType.BYTES29:
    case SchemaType.BYTES30:
    case SchemaType.BYTES31:
    case SchemaType.BYTES32:
    case SchemaType.ADDRESS:
      return pad(hex, { dir: "right", size: staticLength }) as P;
    default:
      return unsupportedStaticField(fieldType);
  }
};

File: packages/services/pkg/protocol-parser/decodeStaticField.go
package protocolparser

import (
	. "latticexyz/mud/packages/services/pkg/schema-type"

	"github.com/andriidski/abiencode-go/convert"
)

func DecodeStaticField(schemaType SchemaType, data string) interface{} {
	if len(data) > 3 && len(data)%2 != 0 {
		panic(ErrInvalidHexLength)
	}

	dataSize := (len(data) - 2) / 2
	if dataSize != StaticAbiTypeToByteLength(schemaType) {
		panic(ErrInvalidHexLengthForStaticField)
	}

	switch schemaType {
	case UINT8:
		return convert.HexToUint8(data)
	case UINT16:
		return convert.HexToUint16(data)
	case UINT24, UINT32:
		return convert.HexToUint32(data)
	case UINT40,
		UINT48,
		UINT56,
		UINT64:
		return convert.HexToUint64(data)
	case UINT72,
		UINT80,
		UINT88,
		UINT96,
		UINT104,
		UINT112,
		UINT120,
		UINT128,
		UINT136,
		UINT144,
		UINT152,
		UINT160,
		UINT168,
		UINT176,
		UINT184,
		UINT192,
		UINT200,
		UINT208,
		UINT216,
		UINT224,
		UINT232,
		UINT240,
		UINT248,
		UINT256:
		return convert.HexToBigInt(data)
	case INT8:
		return convert.HexToInt8(data)
	case INT16:
		return convert.HexToInt16(data)
	case INT24, INT32:
		return convert.HexToInt32(data)
	case INT40,
		INT48,
		INT56,
		INT64:
		return convert.HexToInt64(data)
	case INT72,
		INT80,
		INT88,
		INT96,
		INT104,
		INT112,
		INT120,
		INT128,
		INT136,
		INT144,
		INT152,
		INT160,
		INT168,
		INT176,
		INT184,
		INT192,
		INT200,
		INT208,
		INT216,
		INT224,
		INT232,
		INT240,
		INT248,
		INT256:
		return convert.HexToBigIntSigned(data)
	case BYTES1,
		BYTES2,
		BYTES3,
		BYTES4,
		BYTES5,
		BYTES6,
		BYTES7,
		BYTES8,
		BYTES9,
		BYTES10,
		BYTES11,
		BYTES12,
		BYTES13,
		BYTES14,
		BYTES15,
		BYTES16,
		BYTES17,
		BYTES18,
		BYTES19,
		BYTES20,
		BYTES21,
		BYTES22,
		BYTES23,
		BYTES24,
		BYTES25,
		BYTES26,
		BYTES27,
		BYTES28,
		BYTES29,
		BYTES30,
		BYTES31,
		BYTES32:
		// TODO: should we return a fixed size array per bytes size or a slice?
		return convert.HexToBytes(data)
	case BOOL:
		return convert.HexToBool(data)
	case ADDRESS:
		return convert.HexToAddress(data)
	default:
		panic("unsupported type")
	}
}
